client/allocrunner/csi_hook.go

```diff
@@ -92,8 +92,13 @@ func (c *csiHook) Postrun() error {
 			mode = structs.CSIVolumeClaimWrite
 		}
 
+		source := pair.request.Source
+		if pair.request.PerAlloc {
+			source = source + structs.AllocSuffix(c.alloc.Name)
```

Comment:
I recall an issue where alloc indexes aren't unique, and we may run into two allocations sharing the same id in cases of canaries where multiple deployment versions are running.  Would that cause an issue here?

Reply:
It would! But it's one of our invariants that if you're running a job with `PerAlloc` you can't also use canaries. (It doesn't make sense as a concept with volumes.)

Reply:
Let's make a note of that and/or add a test for that. I fear one day, we relax or change the requirement a bit and miss this assumption in the code.

Reply:
Yup, got it right here: https://github.com/hashicorp/nomad/pull/10136/files#diff-897605edf4370ef2c1cc4c1cdc9ade324caef8450e156f513a2f2aaaa8610e1cR1129

Reply:
I'll definitely add some more commentary to make that clear though.

Reply:
Done in 3b1d19c